core/services/workflows/engine.go

```diff
@@ -397,59 +452,31 @@ func (e *Engine) registerTrigger(ctx context.Context, t *triggerCapability, trig
 	return nil
 }
 
-// loop is the synchronization goroutine for the engine, and is responsible for:
-//   - dispatching new workers up to the limit specified (default = 100)
-//   - starting a new execution when a trigger emits a message on `triggerEvents`
-//   - updating the `executionState` with the outcome of a `step`.
+// stepUpdateLoop is a singleton goroutine per `Execution`, and it updates the `executionState` with the outcome of a `step`.
 //
 // Note: `executionState` is only mutated by this loop directly.
 //
 // This is important to avoid data races, and any accesses of `executionState` by any other
 // goroutine should happen via a `stepRequest` message containing a copy of the latest
 // `executionState`.
-//
-// This works because a worker thread for a given step will only
-// be spun up once all dependent steps have completed (guaranteeing that the state associated
-// with those dependent steps will no longer change). Therefore as long this worker thread only
-// accesses data from dependent states, the data will never be stale.
-func (e *Engine) loop(ctx context.Context) {
+func (e *Engine) stepUpdateLoop(ctx context.Context, executionID string, stepUpdateCh chan store.WorkflowExecutionStep, workflowCreatedAt *time.Time) {
 	defer e.wg.Done()
+	lggr := e.logger.With(eIDKey, executionID)
+	e.logger.Debugf("running stepUpdateLoop for execution %s", executionID)
 	for {
 		select {
 		case <-ctx.Done():
-			e.logger.Debug("shutting down loop")
+			lggr.Debug("shutting down stepUpdateLoop")
 			return
-		case resp, isOpen := <-e.triggerEvents:
-			if !isOpen {
-				e.logger.Error("trigger events channel is no longer open, skipping")
-				continue
-			}
-
-			if resp.Err != nil {
-				e.logger.Errorf("trigger event was an error %v; not executing", resp.Err)
-				continue
-			}
-
-			te := resp.Event
-
-			if te.ID == "" {
-				e.logger.With(tIDKey, te.TriggerType).Error("trigger event ID is empty; not executing")
-				continue
-			}
-
-			executionID, err := generateExecutionID(e.workflow.id, te.ID)
-			if err != nil {
-				e.logger.With(tIDKey, te.ID).Errorf("could not generate execution ID: %v", err)
-				continue
-			}
-
-			err = e.startExecution(ctx, executionID, resp.Event.Outputs)
-			if err != nil {
-				e.logger.With(eIDKey, executionID).Errorf("failed to start execution: %v", err)
+		case stepUpdate, open := <-stepUpdateCh:
+			if !open {
+				lggr.Debug("stepUpdate channel closed, shutting down stepUpdateLoop")
+				return
 			}
-		case stepUpdate := <-e.stepUpdateCh:
 			// Executed synchronously to ensure we correctly schedule subsequent tasks.
-			err := e.handleStepUpdate(ctx, stepUpdate)
+			e.logger.With(eIDKey, stepUpdate.ExecutionID, sRKey, stepUpdate.Ref).
+				Debugf("received step update for execution %s", stepUpdate.ExecutionID)
```

Comment:
This style is less efficient, because it constructs a whole new logger each time. Alternatively:
```suggestion
			e.logger.Debugw(fmt.Sprintf("received step update for execution %s", stepUpdate.ExecutionID), eIDKey, stepUpdate.ExecutionID, sRKey, stepUpdate.Ref)
				
```